helix-core/src/main/java/org/apache/helix/manager/zk/CallbackEventTPFactory.java

```diff
@@ -0,0 +1,76 @@
+package org.apache.helix.manager.zk;
+
+
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+// returns a thread pool object given a HelixManager identifier
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+
+public class CallbackEventTPFactory {
```

Comment:
Just curious, what is the benefit of having this additional layer of the factory? Considering the threadpool is one-one mapping to the executor, the factory does not help with any additional sharing or reusing logic. However, it introduces considerable cleanup overhead. Note that I didn't see any logic in your current version to cleanup the threadpools.

Reply:
TFTR. The threadpool is one-one mapping to HelixManager, not executor. I think having a factory will provide a layer of abstraction for thread pool creation, so each CallbackEventExecutor don't need to know if the thread pool for the helixManager is already created. CallbackEventExecutor could just call `getOrCreateThreadPool ` in its constructor.

Reply:
I see. Makes sense. Thanks for the explanation.